### PR TITLE
CommandHandler to use MaybePromise<T> for execute and isEnabled

### PR DESCRIPTION
**Pull request by amiramw**
This is required to allow async operations in those methods.
https://github.com/theia-ide/theia/issues/435

### CODE REVIEW

File: packages/core/src/browser/keybinding.ts [hunk 1]
```diff
@@ -492,7 +493,7 @@ export class KeybindingRegistry {
     /**
      * Run the command matching to the given keyboard event.
      */
-    run(event: KeyboardEvent): void {
+    async run(event: KeyboardEvent): Promise<void> {
```
Comment (akosyakov) [inline packages/core/src/browser/keybinding.ts]: I am not sure whether this method can be `async`.

New key events can come while previous are not processed yet. Probable, if we make it async then we should queue events.

cc @svenefftinge

**Comment by akosyakov** [comment]
@amiramw i know we have a GH issue about it, do you have other reasons of doing besides the issue? Like you need it to implement some feature or fix some bugs?

**Comment by amiramw** [comment]
@akosyakov i have use cases where the availability check is async.  It is very easy to get there because single async API call in your `isEnabled` flow makes the whole flow async. 
For example with the current implementation every IO related operation is forced to use the sync version. 

Specifically for my use case, I'm trying to embed legacy editors and commands with theia and they are heavily based on async functionality.

File: packages/core/src/common/command.ts [hunk 2]
```diff
@@ -73,12 +74,12 @@ export interface CommandHandler {
...
      */
     // tslint:disable-next-line:no-any
-    execute(...args: any[]): any;
+    execute(...args: any[]): MaybePromise<any>;
     /**
      * Test whether this handler is enabled (active).
      */
     // tslint:disable-next-line:no-any
-    isEnabled?(...args: any[]): boolean;
+    isEnabled?(...args: any[]): MaybePromise<boolean>;
```
Comment (akosyakov) [inline packages/core/src/common/command.ts]: All clients (even transitive, e.g. via `getActiveHandler`)  should be analyzed and should be changed to handle async `isEnabled`.

For instance, `KeybindingRegistry.isActive` will always return `true` for promises.

**Comment by amiramw** [comment]
@akosyakov i started changing the clients as you suggested but at some point i reach phosphore commands in which isEnabled is synced (correct me if i missed something). I opened an issue there https://github.com/phosphorjs/phosphor/issues/369.

**Comment by amiramw** [comment]
Seems that phosphor command are not going to support async isEnabled. Closing this PR.

**Pull request closed without merging**